Homework_3:

Arthasmethil opened the pull request:
-Refactor tests in Void Page Object design pattern.
-Create page Components
-Create java properties file
-Implement Fluent Page Object design patter

Arthasmethil committed:
Homework_3:
-Refactor tests in Void Page Object design pattern.
-Create page Components
-Create java properties file
-Implement Fluent Page Object design patter

shoggy commented on home-tasks/src/main/java/com/epam/tc/hw03/composite/component/common/HeaderMenu.java:
```diff
@@ -0,0 +1,28 @@
+    public void clickOnItem(String itemText) {
+        for (WebElement item : headerMenu) {
+            if (item.getText().equals(itemText)) {
+                item.click();
```
And break the loop

\+ same for other methods like this one

shoggy commented on home-tasks/src/main/java/com/epam/tc/hw03/composite/component/common/Login.java:
```diff
@@ -0,0 +1,64 @@
+        dropdownLoginCaret.click();
+    }
+
+    public void visibilityOfLoginDropdownMenu() {
```
```suggestion
    public void ensureVisibilityOfLoginDropdownMenu() {
```
To be clear

shoggy commented on home-tasks/src/main/java/com/epam/tc/hw03/composite/component/ex1/Icons.java:
```diff
@@ -0,0 +1,28 @@
+
+    public List<WebElement> getIcons() {
+        return benefitIcons;
+    }
```
This getters can be replaced by `@Getter` from `Lombok`. If you want. This is not mandatory

Arthasmethil committed:
Scope of work:
-Refactor selectors for icons

shoggy commented on home-tasks/src/main/java/com/epam/tc/hw03/composite/component/ex1/IframeWindowComponent.java:
```diff
@@ -0,0 +1,29 @@
+    private List<WebElement> iframe;
+
+    @FindBy(id = "frame-button")
+    private List<WebElement> iframeButton;
```
Here you mix two different pages
It looks not so good to me. I'll provide more thoughts regarding this point later

shoggy commented on home-tasks/src/main/java/com/epam/tc/hw03/composite/component/ex2/Checkbox.java:
```diff
@@ -0,0 +1,57 @@
+        selectWindCheckbox();
+        selectSelenRadio();
+        selectYellowColors();
+    }
```
Method name does not reflect it's action

shoggy commented on home-tasks/src/main/java/com/epam/tc/hw03/composite/component/ex2/Log.java:
```diff
@@ -0,0 +1,62 @@
+            windCheckboxLog(),
+            waterCheckboxLog()
+        );
+    }
```
This method does not what it suppose to based on name
Would be better to return all log entries' texts converted and edited in one place

shoggy commented on home-tasks/src/main/java/com/epam/tc/hw03/composite/component/ex2/ServiceDropdownMenu.java:
```diff
@@ -0,0 +1,20 @@
+import org.openqa.selenium.support.ui.ExpectedConditions;
+
+public class ServiceDropdownMenu extends AbstractBaseComponent {
+    @FindBy(xpath = "//a[contains(text(),'Different elements')]")
```
Please use more direct locator (menu menu-item)

shoggy commented on home-tasks/src/main/java/com/epam/tc/hw03/composite/fluent/component/Checkbox.java:
```diff
@@ -0,0 +1,64 @@
+        selenRadioElement.click();
+        yellowColorElement.click();
+        return new DifferentElementsPageFluent(driver);
+    }
```
please choose different name or move this logic to test layer

shoggy commented on home-tasks/src/main/java/com/epam/tc/hw03/composite/fluent/component/Log.java:
```diff
@@ -0,0 +1,63 @@
+            windCheckboxLog(),
+            waterCheckboxLog()
+        );
+    }
```
Same as previous comment regarding logs

shoggy commented on home-tasks/src/main/java/com/epam/tc/hw03/composite/fluent/component/Login.java:
```diff
@@ -0,0 +1,71 @@
+
+    public Login clickToDropdownLoginWindow() {
+        dropdownLoginCaret.click();
+        return new Login(driver);
```
Why do you create new instance instead of returning existing self one?

shoggy commented on home-tasks/src/main/java/com/epam/tc/hw03/composite/fluent/page/DifferentElementsPageFluent.java:
```diff
@@ -0,0 +1,24 @@
+        return this.checkbox;
+    }
+
+    public Log logger() {
```
Dangerous name :) Usually it refers to automation framework logging system, not system under test element

shoggy commented on home-tasks/src/main/java/com/epam/tc/hw03/composite/property/PropertiesData.java:
```diff
@@ -0,0 +1,37 @@
+            fileInputStream.close();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
```
Please use try-with-resources
And throw some exception from catch statement, because running of test with incorrect configuration (properties are part of configuration) has no value

shoggy commented on home-tasks/src/main/resources/hw3/hw3.properties:
```diff
@@ -0,0 +1,5 @@
+
```
This file looks like part of `test` trunk, not `main`

shoggy commented on home-tasks/src/main/java/com/epam/tc/hw03/composite/component/common/HeaderMenu.java:
```diff
@@ -0,0 +1,28 @@
+            if (item.getText().equals(itemText)) {
+                item.click();
+            }
+        }
```
This code is a good candidate to be moved to parent class like `protected void clickOnItem(final Collection<WebElement> list, final String itemText) {...}`

shoggy commented on home-tasks/src/main/java/com/epam/tc/hw03/composite/fluent/component/Checkbox.java:
```diff
@@ -0,0 +1,64 @@
+
+    public DifferentElementsPageFluent visibilityOfCheckboxes() {
+        wait.until(ExpectedConditions.visibilityOf(containerOfCheckboxes));
+        return new DifferentElementsPageFluent(driver);
```
instead of new instance creation you can pass instance of parent DifferentElementPageFluent to constructor and return its value here

Arthasmethil committed:
Scope of homework3:
-Refactor methods
-Lombok is added
-Properties is moved

shoggy commented on home-tasks/src/main/java/com/epam/tc/hw03/composite/component/AbstractBaseComponent.java:
```diff
@@ -0,0 +1,29 @@
+        for (WebElement item : list) {
+            if (item.getText().equals(itemText)) {
+                item.click();
+            }
```
Would be better to add `break` here

shoggy commented on home-tasks/src/main/java/com/epam/tc/hw03/composite/component/ex2/ServiceDropdownMenu.java:
```diff
@@ -0,0 +1,31 @@
+
+    public void clickOnItem(String itemText) {
+        try {
+            super.clickOnItem(getDropdownServiceMenu(), itemText);
```
```suggestion
            clickOnItem(dropdownServiceMenu, itemText);
```

shoggy commented on home-tasks/src/main/java/com/epam/tc/hw03/composite/component/ex2/ServiceDropdownMenu.java:
```diff
@@ -0,0 +1,31 @@
+        try {
+            super.clickOnItem(getDropdownServiceMenu(), itemText);
+        } catch (StaleElementReferenceException e) {
+            super.clickOnItem(getDropdownServiceMenu(), itemText);
```
You have StaleElement because after clicking menu item method continues to iterate over menu items, but another page is already opened having brand new DOM
This should go away after adding `break` in parental method

shoggy commented on home-tasks/src/main/java/com/epam/tc/hw03/composite/support/PropertiesProvider.java:
```diff
@@ -0,0 +1,25 @@
+import lombok.SneakyThrows;
+
+public class PropertiesProvider {
+    private static final String PATH_TO_PROPERTIES = "src/test/resources/hw3/hw3.properties";
+    private static Properties properties;
+
+    @SneakyThrows
+    public PropertiesProvider() {
+        try (FileInputStream fileInputStream = new FileInputStream(PATH_TO_PROPERTIES)) {
+            properties = new Properties();
+            properties.load(fileInputStream);
+        } catch (FileNotFoundException e) {
+            throw new FileNotFoundException("Incorrect file path to properties");
+        }
+    }
```
```suggestion
    private static final String PATH_TO_PROPERTIES = "hw3/hw3.properties";
    private static Properties properties;

    @SneakyThrows
    public PropertiesProvider() {
        try (InputStream fileInputStream = getClass().getClassLoader().getResourceAsStream(PATH_TO_PROPERTIES)) {
            properties = new Properties();
            properties.load(fileInputStream);
        } catch (FileNotFoundException e) {
            throw new FileNotFoundException("Incorrect file path to properties: " + PATH_TO_PROPERTIES);
        }
    }
```

Arthasmethil committed:
Scope of homework4:
-Refactor clickOnItem in AbstractBaseComponent
-Refactor PropertiesProvider
-StaleElement is solved
